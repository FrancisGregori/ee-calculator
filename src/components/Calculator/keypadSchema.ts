import { IKeypadButton } from 'main/types/keypadButton'

const keypadSchema: IKeypadButton[] = [
  {
    gridArea: 'clear',
    value: 'AC',
    color: '#FFA048',
  },
  {
    gridArea: 'divide',
    value: '/',
    color: '#48BDFF',
  },
  {
    gridArea: 'multiply',
    value: '*',
    color: '#48BDFF',
  },
  {
    gridArea: 'minus',
    value: '-',
    color: '#48BDFF',
  },
  {
    gridArea: 'seven',
    value: '7',
  },
  {
    gridArea: 'eight',
    value: '8',
  },
  {
    gridArea: 'nine',
    value: '9',
  },
  {
    gridArea: 'plus',
    value: '+',
    color: '#48BDFF',
  },
  {
    gridArea: 'four',
    value: '4',
  },
  {
    gridArea: 'five',
    value: '5',
  },
  {
    gridArea: 'six',
    value: '6',
  },
  {
    gridArea: 'equal',
    value: '=',
    color: '#FFA048',
  },
  {
    gridArea: 'one',
    value: '1',
  },
  {
    gridArea: 'two',
    value: '2',
  },
  {
    gridArea: 'three',
    value: '3',
  },
  {
    gridArea: 'zero',
    value: '0',
  },
  {
    gridArea: 'dot',
    value: '.',
  },
]

export default keypadSchema
