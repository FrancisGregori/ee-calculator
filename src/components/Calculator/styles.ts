import { ReactComponent as LogoIcon } from 'main/assets/logo.svg'
import styled, { css } from 'styled-components'

export const Container = styled.div`
  width: 100%;
  height: 100%;
  min-height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  padding: 32px;
`

export const Header = styled.header`
  width: 100%;
  max-width: 376px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
  margin-bottom: 48px;
  @media (max-width: 480px) {
    max-width: 248px;
  }
`
export const Logo = styled(LogoIcon)`
  max-height: 40px;
  margin-bottom: 16px;
`

export const Title = styled.h1`
  width: 100%;
  color: #fff;
  text-transform: uppercase;
  font-size: 24px;
  border-top: 1px solid #fff;
  padding-top: 16px;
  line-height: 1;
`
export const Content = styled.div`
  display: flex;
  height: auto;
  align-items: center;
  flex-direction: column;
  background: #212121;
  box-shadow: 0 0 16px rgba(0, 0, 0, 0.25);
  border-radius: 16px;
  overflow: hidden;
  padding: 24px;
  max-width: 376px;
  @media (max-width: 480px) {
    padding: 16px;
  }
`

export const Display = styled.div`
  width: 100%;
  max-width: 328px;
  height: 144px;
  display: flex;
  align-items: flex-end;
  justify-content: flex-end;
  padding-bottom: 32px;
  text-align: right;
  flex-direction: column;
  @media (max-width: 480px) {
    padding-bottom: 16px;
    height: 114px;
    max-width: 248px;
  }
`

export const Value = styled.span<{
  ref: any
  size: number
}>`
  max-width: 100%;
  font-style: normal;
  font-weight: 400;
  font-size: 40px;
  color: #fff;
  @media (max-width: 480px) {
    font-size: 32px;
  }
  ${({ size }) =>
    size > 16 &&
    css`
      font-size: 24px;
      @media (max-width: 480px) {
        font-size: 18px;
      }
    `}
  ${({ size }) =>
    size > 54 &&
    css`
      font-size: 18px;
      @media (max-width: 480px) {
        font-size: 14px;
      }
    `}
  ${({ size }) =>
    size > 120 &&
    css`
      font-size: 16px;
      @media (max-width: 480px) {
        font-size: 12px;
      }
    `}
`

export const Calculation = styled.small<{ hide: boolean }>`
  font-style: normal;
  font-weight: 400;
  font-size: 24px;
  color: #717171;
  padding-top: 16px;
  @media (max-width: 480px) {
    font-size: 16px;
  }
  &:after {
    content: ' =';
  }

  ${({ hide }) =>
    hide &&
    css`
      opacity: 0;
    `}
`

export const Keyboard = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr 1fr 1fr 1fr;
  grid-gap: 16px;
  grid-template-areas:
    'clear divide multiply minus'
    'seven eight nine plus'
    'four five six plus'
    'one two three equal'
    'zero zero dot equal';

  @media (max-width: 480px) {
    grid-gap: 8px;
  }
`
