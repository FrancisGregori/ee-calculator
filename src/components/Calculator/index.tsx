import React, { useRef } from 'react'

import Button from 'components/Button'
import { useCalculatorContext } from 'main/context/CalculatorProvider'

import keypadSchema from './keypadSchema'
import {
  Container,
  Header,
  Logo,
  Title,
  Content,
  Display,
  Value,
  Calculation,
  Keyboard,
} from './styles'

const Calculator = () => {
  const valueRef = useRef<HTMLSpanElement>()
  const { displayValue, calculationValue } = useCalculatorContext()

  return (
    <Container>
      <Header>
        <Logo />
        <Title>EE Calculator</Title>
      </Header>
      <Content>
        <Display>
          <Value ref={valueRef} size={calculationValue.length}>
            {calculationValue}
          </Value>
          <Calculation
            hide={displayValue === 0 || !isNaN(Number(calculationValue))}
          >
            {displayValue}
          </Calculation>
        </Display>
        <Keyboard>
          {keypadSchema.map((item) => (
            <Button
              key={item.value}
              gridArea={item.gridArea}
              value={item.value}
            />
          ))}
        </Keyboard>
      </Content>
    </Container>
  )
}

export default Calculator
