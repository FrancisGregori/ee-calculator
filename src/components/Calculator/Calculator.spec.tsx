import React from 'react'

import { render } from '@testing-library/react'

import Calculator from './index'

describe('Calculator component', () => {
  it('should render correctly', async () => {
    const { asFragment } = render(<Calculator />)

    expect(asFragment()).toMatchSnapshot()
  })
})
