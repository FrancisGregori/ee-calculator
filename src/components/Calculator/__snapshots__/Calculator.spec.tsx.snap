// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnTZi kPCwpg"
    data-testid="calculator"
  >
    <header
      class="sc-dkzDqf lhSXUY"
    >
      <svg
        class="sc-hKMtZM fsDArr"
      />
      <h1
        class="sc-eCYdqJ tMyMz"
      >
        EE Calculator
      </h1>
    </header>
    <div
      class="sc-jSMfEi gWjAWg"
    >
      <div
        class="sc-gKXOVf gaUIPx"
      >
        <span
          class="sc-iBkjds fzXjGf"
        />
        <small
          class="sc-ftvSup iFhWjw"
        >
          0
        </small>
      </div>
      <div
        class="sc-papXJ hmXwZV"
      >
        <button
          class="sc-bczRLJ gQUVyn"
        >
          AC
        </button>
        <button
          class="sc-bczRLJ ipyojJ"
        >
          ÷
        </button>
        <button
          class="sc-bczRLJ jlezhS"
        >
          ×
        </button>
        <button
          class="sc-bczRLJ crdNkW"
        >
          −
        </button>
        <button
          class="sc-bczRLJ kRLmJt"
        >
          7
        </button>
        <button
          class="sc-bczRLJ dUMCed"
        >
          8
        </button>
        <button
          class="sc-bczRLJ gordFS"
        >
          9
        </button>
        <button
          class="sc-bczRLJ keLDUZ"
        >
          +
        </button>
        <button
          class="sc-bczRLJ jUkKws"
        >
          4
        </button>
        <button
          class="sc-bczRLJ kEhILu"
        >
          5
        </button>
        <button
          class="sc-bczRLJ XvNRA"
        >
          6
        </button>
        <button
          class="sc-bczRLJ fcrgRr"
        >
          =
        </button>
        <button
          class="sc-bczRLJ HwUuW"
        >
          1
        </button>
        <button
          class="sc-bczRLJ dXsgLK"
        >
          2
        </button>
        <button
          class="sc-bczRLJ gdxaA"
        >
          3
        </button>
        <button
          class="sc-bczRLJ jzQVjv"
        >
          0
        </button>
        <button
          class="sc-bczRLJ hplbxF"
        >
          .
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
